better-code/src/chapter-2-contracts.md

```diff
@@ -0,0 +1,1129 @@
+  labeled unsafe implicitly promise not to have arbitrary “undefined
+  behavior” when their preconditions are violated.  But that's a
+  blanket guarantee that you don't spell out in contracts.
+
+And if you ever find yourself with something that's clearly a bug but
+you can't decide which code to blame, that's a sign that a contract is
+missing or incomplete.  A system that exposes these situations, where
+everybody is playing by the rules but things still go wrong, is known
+by the technical term “footgun.”
+
+### Type Invariants
+
+The other contribution of Meyer's Design by Contract was to apply the
+idea of “invariants” to user-defined types with private parts.  A
+*class invariant* (or *type invariant*), is a condition that holds at
+a type's public API boundary—whenever a type interacts with its
+clients.  When we talk about an instance being “in a good state,” we
+mean that its type's invariants are satisfied.
+
+My favorite example is this type whose public interface is like an
+array of pairs, but stores the elements of those pairs in a pair of
+arrays.[^array-pairs]
+
+[^array-pairs]: You might want to use a type like this one to store
+    `(Bool, Int64)` pairs without wasting storage for padding, or to
+    leverage SIMD operations.
+
+```swift
+/// A series of `(X, Y)` pairs.
+struct PairArray<X, Y> {
+  // Invariant: `xs.count == ys.count`
+
+  /// The first part of each element.
+  private var xs: [X] = []
+
+  /// The second part of each element.
+  private var ys: [Y] = []
+
+  /// An empty instance.
+  public init() {}
+
+  /// The `i`th element.
+  public subscript(i: Int) -> (X, Y) {
+    (xs[i], ys[i])
+  }
+
+  /// The length.
+  public var count: Int { ys.count }
+
+  /// Adds `e` to the end.
+  public mutating func append(_ e: (X, Y)) {
+    xs.append(e.0)
+    ys.append(e.1)
+  }
+}
+```
+
+The invariant for this type is that the private arrays have the same
+length.  It's important to remember that invariants only hold at a
+type's public interface boundary and are routinely violated,
+temporarily, durign a mutation.  For example, in `append`, we have to
+grow one of the arrays first, which breaks the invariant until we've
+done the other push_back.  That's not a problem because the arrays are
+private—that “bad” state is *encapsulated* by the type, and
+invisible to clients.  By the time we return from `append`, everything
+is back in order.
+
+Because we can control access to the visibility of “bad” states, there
+is a formula for upholding a type invariant.  We
+
+- make stored properties `private` or `private(set)`
+- establish the invariant in the type's `init` method(s), and
+- in `mutating` methods and in the mutation clauses of subscripts
+  and computed properties, ensure the invariant is intact:
+  - before returning to the caller
+  - before passing `self` as a parameter
+  - before calling a non-private method or subscript—or accessing a non-private computed property—of `self`.
```

Comment:
These are obviously stated in Swift-specific terminology. I wonder if some acknowledgement that other languages have equivalent constructs would be useful. Related: `private(set)` is mostly meaningless to a non-Swift programmer, and I wonder if it's adding any real value here. I assume it was added for completeness, but that completeness might be coming at the expense of clarity for users of other languages.

Reply:
> I wonder if some acknowledgement that other languages have equivalent constructs would be useful.

Good question.  But this will come up again, and I don't think we want to repeat it throughout the book.  I suggest you look at the intro chapter and see if there's an appropriate way to express what you mean that fits.  If not, maybe it's not that important.

Yes, `private(set)` is for completeness; we are giving a complete list of rules.  

Swift isn't the end.  IMO we are going to have to announce a Rust version of this book very shortly after it comes out.

Reply:
>  I suggest you look at the intro chapter

Good suggestion - can do.